packages/saas-ui-auth/src/services/supabase.ts

```diff
@@ -6,103 +6,190 @@ import {
   AuthProviderProps,
 } from '../'
 
+// I think this should be added as peerDep
+import type {
+  AuthChangeEvent,
+  AuthResponse,
+  OAuthResponse,
+  Provider,
+  Session,
+  SupabaseClient,
+  VerifyEmailOtpParams,
+  VerifyMobileOtpParams,
+} from "@supabase/supabase-js";
+
 interface RecoveryParams {
-  access_token?: string
-  refresh_token?: string
-  expires_in?: string
-  token_type?: string
-  type?: string
+  access_token?: string;
+  refresh_token?: string;
+  expires_in?: string;
+  token_type?: string;
+  type?: string;
 }
 
 const getParams = (): RecoveryParams => {
-  const hash = window.location.hash.replace('#', '')
-  return hash.split('&').reduce<any>((memo, part) => {
-    const [key, value] = part.split('=')
-    memo[key] = value
-    return memo
-  }, {})
-}
-
-export const createAuthService = (supabase: any): AuthProviderProps => {
+  const hash = window.location.hash.replace("#", "");
+  return hash.split("&").reduce<any>((memo, part) => {
+    const [key, value] = part.split("=");
+    memo[key] = value;
+    return memo;
+  }, {});
+};
+
+export const createAuthService = (
+  supabase: SupabaseClient<any, "public", any>
+): AuthProviderProps => {
   const onLogin = async (params: AuthParams, options?: AuthOptions) => {
-    const { user, error } = await supabase.auth.signIn(
-      params as unknown,
-      options
-    )
-
-    if (user) {
-      return user as User
-    } else if (error) {
-      throw error
+    function authenticate() {
+      const { email, password, provider, phone } = params;
+      if (email && password) {
+        return supabase.auth.signInWithPassword({
+          email,
+          password,
+        });
+      } else if (email) {
+        return supabase.auth.signInWithOtp({ email });
+      } else if (provider) {
+        return supabase.auth.signInWithOAuth({
+          provider: provider as Provider,
+        });
+      } else if (phone && password) {
+        return supabase.auth.signInWithPassword({ phone, password });
+      } else if (phone) {
+        return supabase.auth.signInWithOtp({ phone });
+      }
+      throw new Error("Could not find correct authentication method");
+    }
+    const resp = await authenticate();
+
+    if (resp.error) {
+      throw resp.error;
+    }
+    if (isOauthResponse(resp)) {
+      const userResp = await supabase.auth.getUser();
+      if (userResp.error) {
+        throw userResp.error;
+      }
+      return userResp.data.user;
     }
-  }
+    return resp.data.user;
+  };
 
   const onSignup = async (params: AuthParams, options?: AuthOptions) => {
-    const { email, password, ...data } = params
-    const { user, error } = await supabase.auth.signUp(
-      {
-        email,
-        password,
-      },
-      {
+    const { email, password, phone, captchaToken, emailRedirectTo, data } =
+      params;
+    if (!password) {
```

Comment:
I don't think we need to validate here, just let the supabase client throw an error, but if it's required it should be required in the params type for signup.

Validation is handled in the Auth forms.

Reply:
I just went down that route. If we don't validate this way in runtime, we have to properly represent everything exactly in the types. This would lead to basically implementing Supabase API for all providers.

To be honest, I think the runtime checks are cleaner for now. If adding more auth providers, we can thing about better types.